File: web/src/features/auth/authApi.ts
import { apiClient } from '@/services/apiClient';
import type { LoginFormValues } from './authSchema';
import type { RegisterFormValues } from './registerSchema';

export interface AuthUser {
  id: number;
  name: string;
  email: string;
  role: 'CUSTOMER' | 'PROVIDER' | 'ADMIN';
}

interface AuthResponse {
  success: boolean;
  data: {
    token: string;
    user: AuthUser;
  };
}

export async function login(values: LoginFormValues) {
  const { data } = await apiClient.post<AuthResponse>('/auth/login', values);
  return data.data;
}

export async function register(values: RegisterFormValues) {
  const { data } = await apiClient.post<AuthResponse>('/auth/register', values);
  return data.data;
}

export async function getCurrentUser() {
  const { data } = await apiClient.get<{ success: boolean; data: AuthUser }>('/auth/me');
  return data.data;
}

export async function changePassword(values: {
  currentPassword: string;
  newPassword: string;
}) {
  await apiClient.patch<{ success: boolean; data: { message: string } }>(
    '/auth/change-password',
    values,
  );
}

// Always resolves on 200 — the backend gives the exact same generic
// response whether or not the email belongs to a real account, so there
// is nothing here to distinguish.
export async function requestPasswordReset(email: string): Promise<void> {
  await apiClient.post<{ success: boolean; data: { message: string } }>(
    '/auth/forgot-password',
    { email },
  );
}

// Rejects (via apiClient's normal error handling) for an invalid/expired/
// already-used token or a too-short password — the caller surfaces
// whatever message the backend sends back.
export async function resetPassword(token: string, newPassword: string): Promise<void> {
  await apiClient.post<{ success: boolean; data: { message: string } }>(
    '/auth/reset-password',
    { token, newPassword },
  );
}

File: mobile/lib/features/auth/data/auth_api.dart
import '../../../core/network/api_client.dart';

/// Auth endpoints. The token is attached by the Dio interceptor, so nothing
/// here passes one by hand the way the old `auth_api.dart` did.
class AuthApi {
  const AuthApi(this._client);

  final ApiClient _client;

  Future<Map<String, dynamic>> login({
    required String email,
    required String password,
  }) async => Map<String, dynamic>.from(
    await _client.post(
          '/auth/login',
          body: {'email': email, 'password': password},
        )
        as Map,
  );

  Future<Map<String, dynamic>> register(Map<String, dynamic> payload) async =>
      Map<String, dynamic>.from(
        await _client.post('/auth/register', body: payload) as Map,
      );

  Future<Map<String, dynamic>> currentUser() async =>
      Map<String, dynamic>.from(await _client.get('/auth/me') as Map);

  /// Only name/phone are ever sent — email, role and password each have
  /// their own separate path and are never accepted here (see the
  /// backend's auth.service.js updateCurrentUser doc comment). Passing an
  /// empty string for `phone` clears the stored number (the backend treats
  /// a blank value as "clear"); passing Dart `null` omits the field
  /// entirely, leaving it untouched.
  Future<Map<String, dynamic>> updateProfile({
    String? name,
    String? phone,
  }) async => Map<String, dynamic>.from(
    await _client.patch(
          '/auth/me',
          body: {'name': ?name, 'phone': ?phone},
        )
        as Map,
  );

  Future<void> changePassword({
    required String currentPassword,
    required String newPassword,
  }) async {
    await _client.patch(
      '/auth/change-password',
      body: {'currentPassword': currentPassword, 'newPassword': newPassword},
    );
  }

  /// Always succeeds the same way whether or not the email belongs to a
  /// real account — the backend's response never reveals which.
  Future<void> requestPasswordReset(String email) async {
    await _client.post('/auth/forgot-password', body: {'email': email});
  }
}
